docs/how-to-guides/add-a-custom-ros-message.md

```diff
@@ -0,0 +1,100 @@
+# Add a custom ROS message
+
+## Description
+
+During the Autoware development, you will probably need to define your own messages, This page lists a simple example of adding a custom message to Autoware, For the general ROS2 tutorial see [Create custom msg and srv files](http://docs.ros.org/en/galactic/Tutorials/Beginner-Client-Libraries/Custom-ROS2-Interfaces.html).
+
+## How to create custom message
+
+Make sure you are in the Autoware workspace, and then run the following command to create a new package.
+
+For example we create a package to define sensor messages.
+
+1. Create a package
+
+   ```console
+   ros2 pkg create --build-type ament_cmake autoware_sensing_msgs
+   ```
+
+1. Create custom messages
+
+   You should create `.msg` files and place them in the `msg` directory.
+
+   **NOTE**: The initial letters of the `.msg` and `.srv` files must be capitalized.
+
+   For example we make msg files `GnssInsOrientation.msg` and `GnssInsOrientationStamped.msg` to define GNSS INS orientation messages:
+
+   ```console
+   mkdir msg
+   cd msg
+   touch GnssInsOrientation.msg
+   touch GnssInsOrientationStamped.msg
+   ```
+
+   The `GnssInsOrientation.msg` with the following content:
+
+   ```c++
+   geometry_msgs/Quaternion orientation
+   float32 rmse_rotation_x
+   float32 rmse_rotation_x
+   float32 rmse_rotation_x
+   ```
+
+   In this case, the custom message uses a message from another message package `geometry_msgs/Quaternion`.
+
+   The `GnssInsOrientationStamped.msg` with the following content:
+
+   ```c++
+   std_msgs/Header header
+   GnssInsOrientation orientation
+   ```
+
+   In this case, the custom message uses a message from another message package `std_msgs/Header`.
+
+1. CmakeList.txt
+
+   In order to use this custom message in `C++` or `Python` languages, we need add the following lines to `CmakeList.txt`:
+
+   ```cmake
+   set(msg_files
+   "msg/GnssInsOrientation.msg"
+   "msg/GnssInsOrientationStamped.msg")
+   set(msg_dependencies
+   geometry_msgs
+   std_msgs)
+   rosidl_generate_interfaces(${PROJECT_NAME}
+   ${msg_files}
+   DEPENDENCIES ${msg_dependencies}
+   ADD_LINTER_TESTS)
+   ```
+
+1. package.xml
+
+   We need to declare relevant dependencies in `package.xml`. For the above example we need to add the following content:
+
+   ```xml
+   <depend>geometry_msgs</depend>
+   <depend>std_msgs</depend>
+   <buildtool_depend>rosidl_default_generators</buildtool_depend>
+   <exec_depend>rosidl_default_runtime</exec_depend>
+   <member_of_group>rosidl_interface_packages</member_of_group>
```

Comment:
How about this? To be consistent with the existing packages.

```suggestion
  <buildtool_depend>ament_cmake_auto</buildtool_depend>
  <buildtool_depend>rosidl_default_generators</buildtool_depend>

  <depend>geometry_msgs</depend>
  <depend>std_msgs</depend>

  <exec_depend>rosidl_default_runtime</exec_depend>

  <test_depend>ament_lint_auto</test_depend>
  <test_depend>ament_lint_common</test_depend>

  <member_of_group>rosidl_interface_packages</member_of_group>
```

Reply:
delete:
```xml
<buildtool_depend>ament_cmake_auto</buildtool_depend>
<test_depend>ament_lint_auto</test_depend>
 <test_depend>ament_lint_common</test_depend>
```
Thanks for your suggestion, but I think this tutorial should only cover the dependencies needed for custom messages and not focus on the basic dependencies needed for ROS packages.

Reply:
It's okay, but I believe those messages should almost always be added.
I'll leave it to you.

Reply:
Good idea, `ament_cmake _auto` is a great tool, I will replace `ament_cmake` with `ament_cmake_auto` in `package.xml` and `CmakeLists.txt`.